improve MallocUsage() accuracy

LarryRuane opened the pull request:
The `MallocUsage()` function takes an allocation size as an argument and returns the amount of physical memory consumed, which is greater due to memory allocator overhead and alignment. It was first added in 2015 (first commit of #6102), but its accuracy has degraded as memory allocation libraries have evolved. It's used when it's important that large data structures, such as the coins cache and mempool, should use a predictable, configurable (limited) amount of physical memory (see the `-dbcache` and `-maxmempool` configuration options), as well as a few other places.

sipa figured out a concise, efficient [expression](https://github.com/bitcoin/bitcoin/pull/27748#discussion_r1215541260) that this function can use, and that's what's implemented here.

Also add a unit test, which is more helpful than usual in this case since platforms, operating systems, and libraries vary significantly in this area.

LarryRuane commented:
The results of the old and new versions of this function should differ only slightly; it would be bad if the new one gave very different results, because node operators might find too much memory being consumed, or (not as bad) not enough. To manually test this, I ran `-reindex-chainstate` on Ubuntu, and watched memory usage by printing `/proc/pid/statm` and watching the `data` (6th) field. On master, the value settled to 315140 (units are 4k pages); with this PR it was 315606, which indicates that the new version produces slightly smaller results, but they're close.

LarryRuane commented:
Changing to draft because CI on a couple of the platforms failed the new test, so the physical memory calculation will need to be different on those platforms, I try to figure that out.

martinus commented:
I can reproduce the same results with glibc 2.31 and 2.38, in 32bit and 64bit. I don't think it needs to be perfect on all platforms, I think your new calculation is fine.

I played a bit with a reproducer, here is mine:  https://godbolt.org/z/zqhzxoobx I refactored your `MallocUsage` formula a bit, but it's basically the same

LarryRuane commented:
Force-pushed to 04f3fbe308728d2fb18fad418603cab97f0b9f59 - rebase only. (CI is still expected to fail, I'm working on that now.)

LarryRuane commented:
@martinus
Can you think of why the macOS and Win64 are failing (I have only Ubuntu, which passes)? As an experiment, I changed the memory usage ratio needed to pass to the resource map being just less than `std_map`, but not even that passes (that is, the resource map uses more space). Both of those platforms fail the same way (you can see this in the CI results, this is Win64):
```
D:/a/bitcoin/bitcoin/src/test/pool_tests.cpp(183): error: in "pool_tests/memusage_test": 
  check memusage::DynamicUsage(resource_map) <= memusage::DynamicUsage(std_map) * 100 / 100 
  has failed [466624 > 451088]
```
On earlier runs, some other tests have failed, but I'm not sure if those are spurious failures.

I'm not sure what to do now, any thoughts or suggestions are welcome!

LarryRuane commented:
@sipa - I don't think the test failures could be due to differences in memory allocation overhead across platforms. The actual memory allocation overhead isn't being tested at all. To do that would probably require a test that uses `/proc/meminfo` (which doesn't exist on all platforms) or similar.

I don't mind giving up and closing this PR, but as @martinus said in an [earlier comment](https://github.com/bitcoin/bitcoin/pull/28531#issuecomment-1818440265), this calculation doesn't need to be perfect, just better.

I think this test failure must be due to differences in `std::unordered_map` implementations. That container does various allocations for different reasons, and it must be that with this PR's change in memory usage calculation, the `std::unordered_map` on macOS and Win64 (at least) are doing more a kind of allocation that the new calculation thinks has higher overhead. Something like that.

I thought I had it figured out; I verified with the debugger that as the test adds the 10k entries to the two maps, it has to repeatedly grow the bucket array. The allocation sizes (in bytes) for the bucket array that I saw here on Ubuntu were: 104, 232, 472, 1026, 2056, 4328, 8872, 18856, 40696, and finally, 82184. The tradeoff one makes when using the pool resource allocator is that memory allocations that are freed but then those same sizes not allocated again, those allocations are, in effect, leaked (until the resource is destroyed). The pool resource works great when freeing and allocating the same sizes (rounded up to the alignment, usually 8 bytes) repeatedly. This is, of course, the case for the individual map entries.

So the latest force-push simply calls `reserve()` initially on both maps (this is only really needed for the pool resource map). Debugging verifies that this does prevent the bucket array from being reallocated, here on Ubuntu. But it didn't fix the problem on at least macOS and Win64. Maybe those platforms' implementations of `std::unordered_map` ignore the `reserve()` hint?

It's difficult when CI fails on platforms that I don't have access to. Do others have a solution to that? Can I install macOS on Ubuntu or Windows 10 (I have both) as a VM? Maybe I can set up my Windows laptop to build and run the debugger, but that seems like a lot of work.

martinus commented:
@LarryRuane The problem here is that the `DynamicUsage(const std::unordered_map<X, Y, Z>& m)` just estimates the memory usage by thinking that internally it uses a node with a single pointer (see `struct unordered_node : private X` in `memusage.h`). That this estimation is probably not correct on all platforms, e.g. windows seems to use a double linked list, so this underestimates memory usage.

The inconcsistency now comes when using `DynamicUsage` with a pool resource. Here the actual real world memory usage is counted (because it uses a custom allocator, there's no need for estimating node size), and here the real world usage is reflected in the number.

So, with your new MallocUsage the numbers are most likely more correct, but this seems to trigger the incorrect underestimation for `DynamicUsage(const std::unordered_map<X, Y, Z>& m)`

LarryRuane commented:
@martinus - thanks, that makes perfect sense! I hadn't thought of the possibility of the nodes having two pointers instead of one (double instead of single linked list). Seems like a bad design, but anyway.

I don't think there would be much benefit to fixing `DynamicUsage(const std::unordered_map<X, Y, Z>& m)` to account for this (make it dependent on platform), because this function is no longer important since we're using the pool resource. It's probably really used only in the test. So it's really just a test problem. I'll see what I can do to fix this.

LarryRuane commented:
**NOTE**, this PR isn't ready to merge; at the very least it has some debug prints that need to be removed.

CI was failing on some platforms (Windows and macOS), in `test/functional/mempool_limit.py: test_mid_package_eviction()`. I think this test is fragile; this very-unrelated PR shouldn't be able to cause it to fail IMO. The problem turned out to be the following:

This test calls the test utility function `fill_mempool()`. This function creates independent physically-large (around 65kb) transactions at varying fee rates until the one with the lowest feerate is evicted due to the mempool becoming full. The functional tests run with `-maxmempool=5` (MB), so that requires about 72 transactions.

After `fill_mempool()` returns, the mempool likely won't be _completely_ full, because these were large transactions, so there's some space left but not enough for another transaction. The test, `test_mid_package_eviction`, then submits a smaller transaction (weight around 8k) called `mempool_evicted_tx` with a fee just barely high enough to be accepted, and expects that transaction to be accepted. (It has this name, `mempool_evicted_tx`, because the test expects it to be evicted later.)

With this PR, that transaction was not accepted on Windows and macOS. Some debug prints show that on those platforms, when `fill_mempool()` returns, there are only 1200 bytes of free space in the mempool. This is because the `std::unordered_map` library implementations are different on these platforms, and this causes the memory usage calculation (for the entire map) to be slightly different in this PR. So it's really just bad luck. This is why I think the test is fragile.

After trying a few fixes to the test that didn't work, what finally worked (the current branch) is, to have `fill_mempool()`, just before it returns, submit another transaction that _replaces_ (RBF) an existing one in the mempool, but is slightly smaller. This frees up about 10kb of additional space, which allows room for the 8k transaction to be reliably accepted. This doesn't break the other tests that call `fill_mempool()`.

This attempt didn't work the first time, however, because the mempools wouldn't reliably sync in `test/functional/p2p_1p1c_network.py`. This test starts 4 nodes, and calls `fill_mempool()` on node 0 then waits for all 4 mempools to become synced. Sometimes they do but other times this never happens. I wonder if there might be an unrelated bug (I can investigate later), but what often happens (it's timing dependent and thus random) is that the replacement tx (added by this PR) does appear on all nodes, but a low-feerate tx that was evicted on node 0 remains on nodes 1-3. I can explain what I'm seeing in more detail later if needed. What ended up fixing this is to [delay](https://github.com/bitcoin/bitcoin/pull/28531/files#diff-1a6267ad805cafba25aba03b9c6e3f8315b3f9eb9a938e7c64d4c2b8fac525f9R91) a few seconds before submitting the RBF transaction. It's ugly, maybe there's a better way, but syncing mempools is kind of non-deterministic anyway, so I don't think it's terribly bad.

@glozow and @theStack, can you let me know your opinion? You two have worked a lot in this area.

LarryRuane commented:
Your patch is much better, @theStack, thanks! It's simpler and doesn't need that weird `sleep`. Force-pushed your patch. If it passes CI, I think this is the way to go. I had tried something similar, but I didn't think to make the sending of that extra (smaller) transaction conditional based on available mempool space.

LarryRuane commented:
CI passed, thanks @theStack. I just force-pushed swapping of the two commits, the test commit has to be first for each commit to pass CI.

l0rinc commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -92,6 +93,14 @@ def send_batch(fee):
+    # txs with just the mempoolminfee without immediate eviction ("mempool full" error)
+    if node.getmempoolinfo()['usage'] >= 4_990_000:
+        ephemeral_miniwallet.send_self_transfer(from_node=node, fee=num_of_batches * (base_fee / 2),
+                                                utxo_to_spend=confirmed_utxos.pop(0), target_weight = 32500 * 4)
```
After https://github.com/bitcoin/bitcoin/pull/30718 this should likely be:
```suggestion
                                                utxo_to_spend=confirmed_utxos.pop(0), target_vsize=32500)
```

l0rinc replied:
This is introduced here incorrectly and fixed in a later commit - as discussed, could you please add it in the same or after the fix commit?

l0rinc commented on src/memusage.h:
```diff
@@ -177,11 +189,16 @@ static inline size_t DynamicUsage(const std::list<X>& l)
 
+// struct unordered_node adds the container overhead to the given structure X,
+// although this varies across library container implementations (platforms).
+// It is believed that some containers use two pointers per node, so this
```
can we add some references to avoid having to rely on beliefs?

l0rinc reviewed (changes requested):
I think this is an important change, ~I'm running a full `reindex-chainstate` to check for any noticeable behavior changes~.

I left a few recommendations:
```patch
diff --git a/src/memusage.h b/src/memusage.h
index 501f7068ca..5855e9d18b 100644
--- a/src/memusage.h
+++ b/src/memusage.h
@@ -24,9 +24,6 @@
 namespace memusage
 {
 
-/** Compute the total memory used by allocating alloc bytes. */
-static size_t MallocUsage(size_t alloc);
-
 /** Dynamic memory usage for built-in types is zero. */
 static inline size_t DynamicUsage(const int8_t& v) { return 0; }
 static inline size_t DynamicUsage(const uint8_t& v) { return 0; }
@@ -48,7 +45,7 @@ template<typename X> static inline size_t DynamicUsage(const X * const &v) { ret
  *  application data structures require more accurate inner accounting, they should
  *  iterate themselves, or use more efficient caching + updating on modification.
  */
-static inline size_t MallocUsage(size_t alloc)
+static constexpr size_t MallocUsage(size_t alloc)
 {
     // There are few if any actual zero-length allocations; when
     // DynamicUsage(std::vector<X>& v) calls this function, and v.capacity() == 0,
@@ -59,20 +56,16 @@ static inline size_t MallocUsage(size_t alloc)
     // library-dependent.
     if (alloc == 0) return 0;
 
-    static constexpr size_t overhead = sizeof(void*);
-#if defined(__arm__)
-    // tested with ARM 32bit
-    static constexpr size_t step = sizeof(void*) * 2;
-    static constexpr size_t min_alloc = 9;
+#if defined(__arm__) || SIZE_MAX == UINT64_MAX
+    constexpr size_t min_alloc{9};
 #else
-    static constexpr size_t step = 16U;
-    static constexpr size_t min_alloc = sizeof(void*) == 8 ? 9 : 0;
+    constexpr size_t min_alloc{0};
 #endif
-    // step should be nonzero and a power of 2
-    static_assert(step > 0);
-    static_assert((step & (step-1)) == 0);
-    // tested with Linux glibc 2.31 and 2.38, 32bit and 64bit
-    return (std::max(min_alloc, alloc) + overhead + (step - 1)) & ~(step - 1);
+    constexpr size_t overhead{sizeof(size_t)};
+    constexpr size_t step{alignof(std::max_align_t)};
+    static_assert(std::has_single_bit(step));
+
+    return (std::max(min_alloc, alloc) + overhead + (step - 1)) & -step;
 }
 
 // STL data structures
diff --git a/src/test/mempool_tests.cpp b/src/test/mempool_tests.cpp
index f0094dce59..9e9ae4d73c 100644
--- a/src/test/mempool_tests.cpp
+++ b/src/test/mempool_tests.cpp
@@ -804,4 +804,13 @@ BOOST_AUTO_TEST_CASE(MempoolAncestryTestsDiamond)
     BOOST_CHECK_EQUAL(descendants, 4ULL);
 }
 
+BOOST_AUTO_TEST_CASE(BitwiseStepSimplificationTest)
+{
+    for (uint64_t step{1}; step != 0; step <<= 1) {
+        std::cout << step << std::endl;
+        BOOST_CHECK(std::has_single_bit(step));
+        BOOST_CHECK_EQUAL((~(step - 1)), -step);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```

the updated https://godbolt.org/z/45EnoP16T shows that the output is the same as currently

l0rinc reviewed (commented):
Some of the usages seem to contain some magical values, tailored to the previous behavior:
* https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L1066
* https://github.com/bitcoin/bitcoin/blob/master/src/test/disconnected_transactions.cpp#L39-L43

Do any of them need adjustments now?

Edit: finished reindexing, it doesn't seem to result in any performance regression or speedup

<details>
<summary>Details</summary>

```
COMMITS="3171d7465c237cb752eb500285b272f00027d97d 7ff32e40735c3379da4a14fca1af029d67d43488"; \
STOP_HEIGHT=888000; DBCACHE=5000; \
C_COMPILER=gcc; CXX_COMPILER=g++; \
git fetch -aq && (for c in $COMMITS; do git log -1 --oneline $c || exit 1; done) && \
hyperfine \
  --export-json "/mnt/my_storage/rdx-${COMMITS// /-}-${STOP_HEIGHT}-${DBCACHE}-${C_COMPILER}.json" \
  --runs 1 \
  --parameter-list COMMIT ${COMMITS// /,} \
  --prepare "killall bitcoind || true; rm -f /mnt/my_storage/BitcoinData/debug.log; git checkout {COMMIT}; git clean -fxd; git reset --hard; cmake -B build -DCMAKE_BUILD_TYPE=Release -DENABLE_WALLET=OFF -DCMAKE_C_COMPILER=$C_COMPILER -DCMAKE_CXX_COMPILER=$CXX_COMPILER && cmake --build build -j$(nproc) --target bitcoind && ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=$STOP_HEIGHT -dbcache=10000 -printtoconsole=0 || true" \
  --cleanup "cp /mnt/my_storage/BitcoinData/debug.log /mnt/my_storage/logs/debug-{COMMIT}-$(date +%s).log || true" \
  "COMPILER=$C_COMPILER COMMIT={COMMIT} ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=$STOP_HEIGHT -dbcache=$DBCACHE -blocksonly -reindex-chainstate -connect=0 -printtoconsole=0"

3171d7465c test: ensure that `fill_mempool` leaves some room in mempool
7ff32e4073 Test fix

Benchmark 1: COMPILER=gcc COMMIT=3171d7465c237cb752eb500285b272f00027d97d ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=888000 -dbcache=5000 -blocksonly -reindex-chainstate -connect=0 -printtoconsole=0
  Time (abs ≡):        13876.186 s               [User: 17593.222 s, System: 731.343 s]
 
Benchmark 2: COMPILER=gcc COMMIT=7ff32e40735c3379da4a14fca1af029d67d43488 ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=888000 -dbcache=5000 -blocksonly -reindex-chainstate -connect=0 -printtoconsole=0
  Time (abs ≡):        13912.878 s               [User: 17549.791 s, System: 709.007 s]
 
Summary
  COMPILER=gcc COMMIT=3171d7465c237cb752eb500285b272f00027d97d ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=888000 -dbcache=5000 -blocksonly -reindex-chainstate -connect=0 -printtoconsole=0 ran
    1.00 times faster than COMPILER=gcc COMMIT=7ff32e40735c3379da4a14fca1af029d67d43488 ./build/bin/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=888000 -dbcache=5000 -blocksonly -reindex-chainstate -connect=0 -printtoconsole=0
```

</details>

LarryRuane commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -92,6 +93,14 @@ def send_batch(fee):
+    # if mempool is almost full (<10k usage bytes left), submit one extra middle-sized tx,
+    # in order to evict a large filler tx and leave some room; this will enable tests to submit
+    # txs with just the mempoolminfee without immediate eviction ("mempool full" error)
+    if node.getmempoolinfo()['usage'] >= 4_990_000:
```
I think it's okay that it's initially dead code because it makes the test less fragile in general. In theory, a different project could cherry-pick just this test-only commit, and later make some other changes that would have caused this test to break (when it's really a false-positive failure), but because of taking this commit, that doesn't happen (the test continues to pass). Or, a similar way to think of it is, the test could have been written this way to begin with (and that would not have seemed strange). But I don't feel strongly about, I can make that change if you'd like.

I'll update the commit comment, good catch.

l0rinc replied:
I would like to avoid having a false sense of security by putting this before the fix, which usually indicates that the desired behavior is reproduced and is retained after this fix - which isn't the case here exactly. And the dead code part also bothers me for a similar reason.

l0rinc replied:
yes, it's fine if it's in the same commit, of course.

LarryRuane replied:
Fixed.

LarryRuane reviewed (commented):
@l0rinc - thank you for the excellent review and great suggestions. I especially like making `MallocUsage()` a `constexpr` function.

I'll force-push a commit taking all of them except moving the first (test-only) commit (I left a comment explaining my reasoning).

l0rinc commented on src/memusage.h:
```diff
@@ -48,19 +48,31 @@ template<typename X> static inline size_t DynamicUsage(const X * const &v) { ret
+    static constexpr size_t overhead = sizeof(void*);
+#if defined(__arm__)
+    // tested with ARM 32bit
+    static constexpr size_t step = sizeof(void*) * 2;
```
I forgot to include this in the full patch, my bad

LarryRuane commented on src/memusage.h:
```diff
@@ -177,23 +176,33 @@ static inline size_t DynamicUsage(const std::list<X>& l)
 {
 private:
     void* ptr;
+    void* ptr2;
```
No, I'll write a detailed comment later today or tomorrow, but it turns out the lack of this second pointer caused the pool test to fail, because on macOS and Windows; they really do have two pointers per node. I was able to determine this with certainty with debug printing (you can see it in some of the force-pushes above). I added a comment just before the start of this `struct unordered_node` definition to explain this second pointer, did you see that? I didn't call it forward and previous because I really don't know that, I only know that there are 16 bytes in this part of the struct (but I am going to assume that there are two pointers).

l0rinc commented on src/test/validation_flush_tests.cpp:
```diff
@@ -30,7 +30,7 @@ BOOST_AUTO_TEST_CASE(getcoinscachesizestate)
     //
     // See also: Coin::DynamicMemoryUsage().
-    constexpr unsigned int COIN_SIZE = is_64_bit ? 80 : 64;
+    constexpr size_t COIN_SIZE{64};
```
This is a surprising change, can you please explain it?
Do we even get to `BOOST_CHECK_EQUAL(view.DynamicMemoryUsage(), is_64_bit ? 32U : 16U);` here - and if so, how come none of the other values need changing?

l0rinc replied:
[I've checked](https://github.com/bitcoin/bitcoin/pull/32244) and most of the test is dead for the [past 3 years](https://github.com/bitcoin/bitcoin/pull/25325), cc @martinus.

Could you update the modified test all the way to revive it and define the desired behavior?

l0rinc commented on src/memusage.h:
```diff
@@ -48,19 +45,21 @@ template<typename X> static inline size_t DynamicUsage(const X * const &v) { ret
+    if (alloc == 0) return 0;
+
+#if defined(__arm__) || SIZE_MAX == UINT64_MAX
+    constexpr size_t min_alloc{9};
```
could we document this number here, it's definitely not obvious

LarryRuane replied:
That constant came from your [comment](#pullrequestreview-2697495377) (which I've adopted into this PR), so I really don't know how to document this value. Where did you get it from? This stuff tends to have magic numbers; I don't think this PR makes it any worse.

l0rinc commented on src/memusage.h:
```diff
@@ -177,23 +176,33 @@ static inline size_t DynamicUsage(const std::list<X>& l)
 static inline size_t DynamicUsage(const std::unordered_map<X, Y, Z>& m)
 {
-    return MallocUsage(sizeof(unordered_node<std::pair<const X, Y> >)) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
+    return MallocUsage(sizeof(unordered_node<std::pair<const X, Y> >)) * m.size() + MallocUsage(2 * sizeof(void*) * m.bucket_count());
```
`2 * sizeof(void*)` refers to `ptr` + `ptr2`, right? Can we extract that to an explanatory constant?

LarryRuane replied:
Yes, and there is a fairly detailed comment on lines 194-195 (just before the `unordered_set` version of `DynamicUsage`), so I think that's sufficient. But if you feel strongly, I'll try to come up with something.

l0rinc commented on src/memusage.h:
```diff
@@ -212,7 +221,7 @@ static inline size_t DynamicUsage(const std::unordered_map<Key,
     size_t usage_resource = estimated_list_node_size * pool_resource->NumAllocatedChunks();
     size_t usage_chunks = MallocUsage(pool_resource->ChunkSizeBytes()) * pool_resource->NumAllocatedChunks();
-    return usage_resource + usage_chunks + MallocUsage(sizeof(void*) * m.bucket_count());
+    return usage_resource + usage_chunks + MallocUsage(2 * sizeof(void*) * m.bucket_count());
```
> likely two pointers

Is there a way to verify and document that explicitly (pointing to doc or code)?

l0rinc commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -57,16 +57,17 @@ def fill_mempool(test_framework, node, *, tx_sync_fun=None):
+    # 1 to create a final middle-sized tx to evict a filler tx and create room (if needed)
     ephemeral_miniwallet = MiniWallet(node, tag_name="fill_mempool_ephemeral_wallet")
-    test_framework.generate(ephemeral_miniwallet, 1 + num_of_batches * tx_batch_size)
+    test_framework.generate(ephemeral_miniwallet, 2 + num_of_batches * tx_batch_size)
```
same, can we extract the `2` to a named constant?

LarryRuane replied:
I think the comment immediately above explains it sufficiently, and the "1" was not a named constant either (so this pull is not making it worse). But if you feel strongly, I can do something. (As someone once said, the hardest part of programming is coming up with good names!)

LarryRuane commented:
The main change in the latest force push is to address the [review comment](#discussion_r2044004503) about `validation_flush_tests.cpp`. I ended up almost completely rewriting the `getcoinscachesizestate` test case; it was kind of a mess. It was far too fragile because it required equality of various values related to memory usage (which is why this test started failing with this PR, even though nothing had actually broken). This is especially difficult when a test must run in both the 32-bit and 64-bit environments. A test like this should be more of a sanity check. 

This test case also tested things outside of the `GetCoinsCacheSizeState()` function, which is what the test case name implies. The new version tests exactly that function, driving it to return its various possible results. It's more focused and easier to understand now, in my opinion.

Also, I squashed the three commits, two of them test-only, down to a single commit, as [suggested above](#discussion_r2003105459); it's still a pretty small commit.

Reviewers, please let me know if the commit message is too wordy. It's almost entirely about the test changes (I preserved the text from what used to be separate commits), because those were the trickiest parts of this PR. But maybe it would be better for some of that explanation to be here in the PR instead of the commit message.

@l0rinc, thank you for your excellent review comments. There were a few suggestions I didn't take (so far, at least); I'll leave comments as replies to yours.

LarryRuane commented:
Rebased to see if that fixes CI failure (which seems unrelated), the branch was previously rebased on Mar 17 2025.

l0rinc commented on src/memusage.h:
```diff
@@ -212,7 +221,9 @@ static inline size_t DynamicUsage(const std::unordered_map<Key,
     size_t usage_chunks = MallocUsage(pool_resource->ChunkSizeBytes()) * pool_resource->NumAllocatedChunks();
-    return usage_resource + usage_chunks + MallocUsage(sizeof(void*) * m.bucket_count());
+    // Empirically, each element of the bucket array has two pointers on some platforms (Windows and macOS).
+    size_t usage_bucket_array = MallocUsage(2 * sizeof(void*) * m.bucket_count());
```
https://github.com/bitcoin/bitcoin/blob/master/src/coins.h#L219 already claims to account for 1-2 pointers, do we need any other change to synchronize it with the current PR?

l0rinc reviewed (changes requested):
Thanks for getting back to this change.
We should synchronize it with `CCoinsMap`, chich already claims to account for an "overhead of 1 or 2 pointers".
Also, the test should pass before the change as well, so it would be helpful to add it as a separate commit before the change to make it easy to see how it behaves before and after the change as well.
The test still contains a lot of repetition, please see my suggestion on how to compact it a bit more.
There are still uncovered parts of the code, I'd feel more comfortable if all modified code parts are exercised by tests.

LarryRuane committed:
improve MallocUsage() accuracy

Co-authored-by: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>
Co-authored-by: Pieter Wuille <pieter@wuille.net>
Co-authored-by: Martin Leitner-Ankerl <martin.ankerl@gmail.com>
Co-authored-by: l0rinc <pap.lorinc@gmail.com>

# Changes to test/functional/test_framework/mempool_util.py:

## fill_mempool() should call sync_mempools() before returning

We saw a case where a test (p2p_1p1c_network.py) called
raise_network_minfee(), which called fill_mempool() using node 0.

Then raise_network_minfee() returned, and the test called rescan_utxos(),
which called getrawmempool() using a different node (node 1) followed by
getrawtransaction() on each returned transaction, and the test asserted
because a transaction was not found.

This was caused by the timing window between the call to getrawmempool()
and fetching the individual transactions; the transactions were still
being propagated on the P2P network. During this window, a transaction
(returned by getrawmempool()) was evicted (the mempool is close to full
during this test), and did not exist in the mempool by the time it was
attempted to be fetched.

It might make more sense for rescan_utxos() to call sync_mempools() just
before calling getrawmempool(), but it can't because rescan_utxos() is
part of the MiniWallet class, which doesn't have access to test_framework
(but that could probably be changed).

## ensure that `fill_mempool` leaves some room in mempool

Without this change, fill_mempool() may leave the mempool very close
to its memory size limit (-maxmempool). This can cause tests to
incorrectly fail when they submit another transaction expecting it
to succeed. Note that without this change, the same test that fails on
one platform may succeed on another, because their memory allocation
accounting algorithms (how they calculate memory usage, that is,
MallocUsage()) may be slightly different.

LarryRuane committed:
temporary commit (don't merge), add MallocUsage tests

This commit is temporary because it makes one of the new tests fail
intentionally so we can gather some information from CI across the
platforms. (second attempt, fixes CI error private field not used)

l0rinc commented on src/test/util_malloc_usage_tests.cpp:
```diff
@@ -0,0 +1,105 @@
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(util_malloc_usage_tests)
```
@LarryRuane, most of these tests will be executed on your own fork as well (except a few Cirrus fuzzers).
You can freely experiment there - see for example my attempts as https://github.com/l0rinc/bitcoin/pull/20

LarryRuane replied:
I didn't know that, that will prevent clutter here with that temporary stuff, thanks! I'll do that from now on (I want to do a few more similar experiments).

l0rinc replied:
yeah, I also have a few experiments where I'm not sure about the state of CI so I push to my local fork first - though the remaining CIs still surprise me sometimes after upstream push :)

l0rinc commented on src/test/validation_flush_tests.cpp:
```diff
@@ -21,115 +21,113 @@ BOOST_AUTO_TEST_CASE(getcoinscachesizestate)
         CoinsCacheSizeState::CRITICAL);
 
-    // Passing non-zero max mempool usage (512 KiB) should allow us more headroom.
+    // Repeat (continuing with the existing cache) but with a non-zero max mempool;
```
I have extracted the suggested test to a separate PR: https://github.com/bitcoin/bitcoin/pull/33021

achow101 commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

Marking as up for grabs.

cc @l0rinc @sipa
